docs/csharp/whats-new/csharp-7-3.md

```diff
@@ -0,0 +1,206 @@
+ms.date: 05/16/2018
+---
+# What's new in C# 7.3
+
+There are two main themes to the C# 7.3 release. One theme provides features that enable safe code to be as performant as unsafe code. The second theme provides incremental improvements to existing features. In addition, new compiler options were added in this release.
+
+The following new features support the theme of better performance for safe code:
+
+- You can access fixed fields without pinning.
+- You can reassign `ref` local variables.
+- You can use initializers on `stackalloc` arrays.
+- You can use `fixed` statements with any type that supports a pattern.
+- You can use additional generic constraints.
+
+The following enhancements were made to existing features:
+
+- You can test `==` and `!=` with tuple types.
+- You can use expression variables in more locations.
+- You may attach attributes to the backing field of auto-implemented properties.
+- Method resolution when arguments differ by `in` has been improved.
+- Overload resolution now has fewer ambiguous cases.
+
+The new compiler options are:
+
+- `-publicsign` to enable Open Source Software (OSS) signing of assemblies.
+- `-pathmap` to provide a mapping for source directories.
+
+The remainder of this article provides details and links to learn more about each of the improvements.
+
+## Enabling more performant safe code
+
+You should be able to write C# code safely that performs as well as unsafe code. Safe code avoids classes of errors, such as buffer overruns, stray pointers, and other memory access errors. These new features expand the capabilities of verifiable safe code. Strive to write more of your code using safe constructs. These features make that easier.
+
+### Indexing `fixed` fields does not require pinning
+
+Consider this struct:
+
+```csharp
+unsafe struct S
+{
+    public fixed int myFixedField[10];
+}
+```
+
+In earlier versions of C#, you needed to pin a variable to access one of the integers that are part of `myFixedField`. Now, the following code compiles in a safe context:
+
+```csharp
+class C
+{
+    static S s = new S();
+
+    unsafe public void M()
+    {
+        int p = s.myFixedField[5];
+    }
+}
+```
+
+The variable `p` accesses one element in `myFixedField`. You don't need to declare a separate `int*` variable. Note that you still need an `unsafe` context. In earlier versions of C#, you need to declare a second fixed pointer:
+
+```csharp
+class C
+{
+    static S s = new S();
+
+    public void M()
+    {
+        fixed (int* ptr = s.myFixedField)
+        {
+            int p = ptr[5];
+        }
+    }
+}
+```
+
+Fore more information, see the article on the [`fixed` statement](../language-reference/keywords/fixed-statement.md).
+
+### `ref` local variables may be reassigned
+
+Now, `ref` locals may be reassigned to refer to different instances after being initialized. The following code now compiles:
+
+```csharp
+ref VeryLargeStruct refLocal = ref veryLargeStruct; // initialization
```

Comment:
There is one more form of ref local re-assignment: `foreach ref`. It was missing from the spec, but it is implemented and shipped.

```C#
        foreach (ref var v in new int[0])
        {
        }
```

Reply:
@jcouv that means we can make such shoot-in-the-leg things now:
```csharp
var _array = new byte[5];
new Random(42).NextBytes(_array);
Span<byte> span = _array;
foreach (ref byte element in span)
{
    element = 4;
    Console.WriteLine(element);
}
Console.WriteLine(string.Join(" ", _array));
Console.ReadLine();
```
That outputs only `4`s. Wow.